PR: https://github.com/odoo/odoo/pull/287396

From: fb63b76eb9e26842c02bf8215fdb7e3736140486
From: Saif (segn)
Date: 2026-09-14 17:56:11

Breaking data model changes scores: del:3 + add:0, change matches:
-    tax_payable_account_id = fields.Many2one(
-    tax_receivable_account_id = fields.Many2one(
-    advance_tax_payment_account_id = fields.Many2one(

Total Changes: 87

[IMP] account,l10n_*: Move closing accounts to return type

Move the tax closing accounts from `account.tax.group` to
`account.return.type` to make their configuration easier to find and avoid
duplicating the same accounts across tax groups.

Remove the accounts set in `account.tax.group-*.csv` files

task-6237033

closes odoo/odoo#287396

Related: odoo/enterprise#130955
Related: odoo/upgrade#11280
Signed-off-by: Hugo Poncelet (hupo) <hupo@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_tax.py
+++ b/addons/account/models/account_tax.py
@@ -32,23 +32,6 @@ class AccountTaxGroup(models.Model):
     name = fields.Char(required=True, translate=True)
     sequence = fields.Integer(default=10)
     company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
-    tax_payable_account_id = fields.Many2one(
-        comodel_name='account.account',
-        check_company=True,
-        domain="[('account_type', '=', 'liability_payable'), ('non_trade', '=', True)]",
-        string='Tax Payable Account',
-        help="Tax current account used as a counterpart to the Tax Closing Entry when in favor of the authorities.")
-    tax_receivable_account_id = fields.Many2one(
-        comodel_name='account.account',
-        check_company=True,
-        domain="[('account_type', '=', 'asset_receivable'), ('non_trade', '=', True)]",
-        string='Tax Receivable Account',
-        help="Tax current account used as a counterpart to the Tax Closing Entry when in favor of the company.")
-    advance_tax_payment_account_id = fields.Many2one(
-        comodel_name='account.account',
-        check_company=True,
-        string='Tax Advance Account',
-        help="Downpayments posted on this account will be considered by the Tax Closing Entry.")
     country_id = fields.Many2one(
         string="Country",
         comodel_name='res.country',
@@ -64,19 +47,6 @@ class AccountTaxGroup(models.Model):
     )
     pos_receipt_label = fields.Char(string='PoS receipt label')
 
-    @api.constrains('tax_payable_account_id', 'tax_receivable_account_id')
-    def _constrains_payable_receivable_account(self):
-        for tax_group in self:
-            if tax_group.tax_payable_account_id and tax_group.tax_payable_account_id.account_type != 'liability_payable':
-                raise UserError(self.env._("You must select a payable account for 'Tax Payable Account'."))
-            if tax_group.tax_receivable_account_id and tax_group.tax_receivable_account_id.account_type != 'asset_receivable':
-                raise UserError(self.env._("You must select a receivable account for 'Tax Receivable Account'."))
-            if (
-                (tax_group.tax_payable_account_id and not tax_group.tax_payable_account_id.non_trade)
-                or (tax_group.tax_receivable_account_id and not tax_group.tax_receivable_account_id.non_trade)
-            ):
-                raise UserError(self.env._("You must use non-trade accounts for tax groups."))
-
     @api.depends('company_id')
     def _compute_country_id(self):
         for group in self:

--- a/addons/account/models/chart_template.py
+++ b/addons/account/models/chart_template.py
@@ -373,10 +373,6 @@ class AccountChartTemplate(models.AbstractModel):
                     if xmlid not in xmlid2tax_group and not force_create:
                         skip_update.add((model_name, xmlid))
                         continue
-                    if xmlid in xmlid2tax_group:
-                        for field_name in ["tax_payable_account_id", "tax_receivable_account_id"]:
-                            if field_name in values and self.ref(values[field_name], raise_if_not_found=False):
-                                values.pop(field_name, None)
 
                 elif model_name == 'account.tax':
                     if xmlid not in xmlid2tax and 'name' not in values:
@@ -981,6 +977,16 @@ class AccountChartTemplate(models.AbstractModel):
         ])
 
     @api.model
+    def _create_foreign_account(self, company, existing_account, additional_label, reconcilable=False):
+        return self.env['account.account'].create({
+            'name': f"{existing_account.name} - {additional_label}",
+            'code': self.env['account.account'].with_company(company)._search_new_account_code(existing_account.code),
+            'account_type': existing_account.account_type,
+            'reconcile': reconcilable or existing_account.reconcile,
+            'non_trade': existing_account.non_trade,
+            'company_ids': [Command.link(company.id)],
+        })
+
     def _instantiate_foreign_taxes(self, country, company):
         """Create and configure foreign taxes from the provided country.
 
@@ -1003,43 +1009,12 @@ class AccountChartTemplate(models.AbstractModel):
         if taxes_in_country:
             return
 
-        def create_foreign_tax_account(existing_account, additional_label, reconcilable=False):
-            new_code = self.env['account.account'].with_company(company)._search_new_account_code(existing_account.code)
-            return self.env['account.account'].create({
-                'name': f"{existing_account.name} - {additional_label}",
-                'code': new_code,
-                'account_type': existing_account.account_type,
-                'reconcile': reconcilable or existing_account.reconcile,
-                'non_trade': existing_account.non_trade,
-                'company_ids': [Command.link(company.id)],
-            })
-
         existing_accounts = {'': None, None: None}  # keeps tracks of the created account by foreign xml_id
         default_company_taxes = company.account_sale_tax_id + company.account_purchase_tax_id
         chart_template_code = self._guess_chart_template(country=country)
         tax_group_data = self._get_chart_template_data(chart_template_code)['account.tax.group']
         tax_data = self._get_chart_template_data(chart_template_code)['account.tax']
 
-        # Populate foreign accounts mapping
-        # Try to create tax group accounts if not mapped
-        field_and_names = (
-            ('tax_payable_account_id', _("Foreign tax account payable (%s)", country.code)),
-            ('tax_receivable_account_id', _("Foreign tax account receivable (%s)", country.code)),
-            ('advance_tax_payment_account_id', _("Foreign tax account advance payment (%s)", country.code)),
-        )
-        for field, account_name in field_and_names:
-            for tax_group in tax_group_data.values():
-                account_template_xml_id = tax_group.get(field)
-                if account_template_xml_id in existing_accounts:
-                    continue
-                local_tax_group = self.env["account.tax.group"].search([
-                    *self.env['account.tax.group']._check_company_domain(company),
-                    ('country_id', '=', company.account_fiscal_country_id.id),
-                    (field, '!=', False),
-                ], limit=1)
-                if local_tax_group:
-                    existing_accounts[account_template_xml_id] = create_foreign_tax_account(local_tax_group[field], account_name).id
-
         # Try to create repartition lines account if not mapped
         for tax_template in tax_data.values():
             for _command, _id, rep_line in tax_template.get('repartition_line_ids', []):
@@ -1071,7 +1046,7 @@ class AccountChartTemplate(models.AbstractModel):
 
                     if similar_repartition_line:
                         local_tax_account = similar_repartition_line.account_id
-                        similar_account_id = create_foreign_tax_account(local_tax_account, _("Foreign tax account (%s)", country.code))
+                        similar_account_id = self._create_foreign_account(company, local_tax_account, _("Foreign tax account (%s)", country.code))
                         existing_accounts[account_template_xml_id] = similar_account_id.id
 
         # Try to create cash basis account if not mapped
@@ -1089,7 +1064,8 @@ class AccountChartTemplate(models.AbstractModel):
                 account_xml_id = tax_template.get('cash_basis_transition_account_id')
                 if account_xml_id not in existing_accounts:
                     if local_cash_basis_tax:
-                        existing_accounts[account_xml_id] = create_foreign_tax_account(
+                        existing_accounts[account_xml_id] = self._create_foreign_account(
+                            company,
                             local_cash_basis_tax.cash_basis_transition_account_id,
                             _("Cash basis transition account"),
                             reconcilable=True,
@@ -1097,7 +1073,7 @@ class AccountChartTemplate(models.AbstractModel):
 
                     elif account_ids := [rep_line['account_id'] for _command, _id, rep_line in tax_template.get('repartition_line_ids', []) if rep_line.get('account_id')]:
                         local_account = self.env['account.account'].browse(existing_accounts[account_ids[0]])
-                        existing_accounts[account_xml_id] = create_foreign_tax_account(local_account, _("Cash basis transition account"), reconcilable=True).id
+                        existing_accounts[account_xml_id] = self._create_foreign_account(company, local_account, _("Cash basis transition account"), reconcilable=True).id
 
                     else:
                         existing_accounts[account_xml_id] = None
@@ -1105,11 +1081,6 @@ class AccountChartTemplate(models.AbstractModel):
         if has_cash_basis:
             company.tax_exigibility = True
 
-        # Assign the account based on the map
-        for field, _account_name in field_and_names:
-            for tax_group in tax_group_data.values():
-                tax_group[field] = existing_accounts.get(tax_group.get(field))
-
         for tax_template in tax_data.values():
             # This is required because the country isn't provided directly by the template
             tax_template['country_id'] = country.id
